File: Igatha/ContentViewModel.swift
//
//  ViewModel.swift
//  Igatha
//
//  Created by Nizar Mahmoud on 14/10/2024.
//

import Foundation

class ContentViewModel: ObservableObject {
    @Published public var isSOSAvailable: Bool = false
    @Published public var isSOSActive: Bool = false
    
    @Published public var activeAlert: AlertType? = nil
    
    @Published private var devicesMap: [String: Device] = [:]
    public var devices: [Device] {
        return Array(devicesMap.values)
    }
    
    private let emergencyManager: EmergencyManager
    private let proximityScanner: ProximityScanner
    
    init() {
        emergencyManager = EmergencyManager.shared
        proximityScanner = ProximityScanner()
        
        emergencyManager.delegate = self
        proximityScanner.delegate = self
    }
    
    deinit {
        proximityScanner.stopScanning()
    }
    
    func startSOS() {
        emergencyManager.startSOS()
    }
    
    func stopSOS() {
        emergencyManager.stopSOS()
    }
}

extension ContentViewModel: EmergencyManagerDelegate {
    func incidentDetected() {
        DispatchQueue.main.async {
            self.activeAlert = .incidentDetected
        }
    }
    
    func distressSignalStarted() {
        DispatchQueue.main.async {
            self.isSOSActive = true
        }
    }
    
    func distressSignalStopped() {
        DispatchQueue.main.async {
            self.isSOSActive = false
        }
    }
    
    func emergencyManagerAvailabilityUpdate(_ isAvailable: Bool) {
        DispatchQueue.main.async {
            self.isSOSAvailable = isAvailable
        }
    }
}

extension ContentViewModel: ProximityScannerDelegate {
    func scannedDevice(_ device: Device) {
        DispatchQueue.main.async {
            if !self.devicesMap.keys.contains(device.id) {
                self.devicesMap[device.id] = device
            }
            
            self.devicesMap[device.id]?.update(rssi: device.rssi)
        }
    }
    
    func scannerAvailabilityUpdate(_ isAvailable: Bool) {
        guard isAvailable else {
            proximityScanner.stopScanning()
            return
        }
        
        proximityScanner.startScanning()
    }
}

enum AlertType: Identifiable {
    case sosConfirmation
    case incidentDetected
    
    var id: Int {
        switch self {
        case .sosConfirmation:
            return 1
        case .incidentDetected:
            return 2
        }
    }
}

File: Android/app/src/main/java/com/nizarmah/igatha/viewmodel/ContentViewModel.kt
package com.nizarmah.igatha.viewmodel

import android.app.Application
import android.content.Intent
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.nizarmah.igatha.Constants
import com.nizarmah.igatha.model.Device
import com.nizarmah.igatha.service.DisasterDetectionService
import com.nizarmah.igatha.service.EmergencyManager
import com.nizarmah.igatha.service.ProximityScanner
import com.nizarmah.igatha.service.SOSService
import com.nizarmah.igatha.util.PermissionsManager
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

class ContentViewModel(app: Application) : AndroidViewModel(app) {

    private val emergencyManager = EmergencyManager.getInstance(getApplication())
    private val proximityScanner = ProximityScanner(getApplication())

    // SOS state
    val isSOSAvailable: StateFlow<Boolean> = emergencyManager.isSOSAvailable
    val isSOSActive: StateFlow<Boolean> = emergencyManager.isSOSActive

    // ProximityScanner state
    val isProximityScanAvailable: StateFlow<Boolean> = combine(
        proximityScanner.isAvailable,
        PermissionsManager.proximityScanPermitted
    ) { isAvailable, isPermitted ->
        isAvailable && isPermitted
    }.stateIn(viewModelScope, SharingStarted.Eagerly, false)

    // Active alert state
    private val _activeAlert = MutableStateFlow<AlertType?>(null)
    val activeAlert: StateFlow<AlertType?> = _activeAlert.asStateFlow()

    private val _devicesMap = MutableStateFlow<Map<String, Device>>(emptyMap())
    val devices: StateFlow<List<Device>> = _devicesMap.asStateFlow()
        .map { devicesMap ->
            devicesMap.values.sortedByDescending { it.rssi }
        }
        .stateIn(
            viewModelScope,
            SharingStarted.Eagerly,
            emptyList()
        )

    init {
        // Observe the disaster detection availability
        viewModelScope.launch {
            emergencyManager.isDetectorAvailable.collect { available ->
                if (available) {
                    startDisasterDetectionService()
                } else {
                    stopDisasterDetectionService()
                }
            }
        }

        // Observe proximity scanner's scanned devices
        viewModelScope.launch {
            proximityScanner.scannedDevices.collect { device ->
                device?.let { updateDevice(it) }
            }
        }

        // Start or stop the scanner based on availability
        viewModelScope.launch {
            isProximityScanAvailable.collect { available ->
                if (available) {
                    proximityScanner.startScanning()
                } else {
                    proximityScanner.stopScanning()
                }
            }
        }
    }

    private fun startDisasterDetectionService() {
        val context = getApplication<Application>()
        val serviceIntent = Intent(context, DisasterDetectionService::class.java)
        context.startForegroundService(serviceIntent)
    }

    private fun stopDisasterDetectionService() {
        val context = getApplication<Application>()
        val serviceIntent = Intent(context, DisasterDetectionService::class.java)
        context.stopService(serviceIntent)
    }

    private fun startSOSService() {
        val context = getApplication<Application>()
        val serviceIntent = Intent(context, SOSService::class.java)
        context.startForegroundService(serviceIntent)
    }

    private fun stopSOSService() {
        val context = getApplication<Application>()
        val serviceIntent = Intent(context, SOSService::class.java).apply {
            action = Constants.ACTION_STOP_SOS
        }
        context.startService(serviceIntent)
    }

    fun startSOS() {
        if (!isSOSAvailable.value || isSOSActive.value) {
            return
        }

        startSOSService()
    }

    fun stopSOS() {
        stopSOSService()
    }

    fun showSOSConfirmation() {
        if (!isSOSAvailable.value || isSOSActive.value) {
            return
        }

        _activeAlert.value = AlertType.SOSConfirmation
    }

    fun dismissAlert() {
        _activeAlert.value = null
    }

    private fun updateDevice(device: Device) {
        _devicesMap.update { currentMap ->
            val updatedMap = currentMap.toMutableMap()
            updatedMap[device.id] = device
            updatedMap
        }
    }

    override fun onCleared() {
        super.onCleared()
        proximityScanner.deinit()
    }
}

sealed class AlertType(id: Int) {
    object SOSConfirmation : AlertType(1)
}
